fdbserver/VersionedBTree.actor.cpp

```diff
@@ -6100,48 +6108,59 @@ class VersionedBTree {
-				             mBegin.mutation().toString().c_str());
+			constexpr int maxHeightAllowed = 8;
 
+			while (mBegin != mEnd) {
 				// Apply the change to the mutation buffer start boundary key only if
-				//   - there actually is a change (whether a set or a clear, old records are to be removed)
+				//   - there actually is a change (clear or set to new value)
 				//   - either this is not the first boundary or it is but its key matches our lower bound key
 				bool applyBoundaryChange = mBegin.mutation().boundaryChanged &&
 				                           (!firstMutationBoundary || mBegin.key() == update->subtreeLowerBound.key);
+				bool boundaryExists = cursor.valid() && cursor.get().key == mBegin.key();
+
+				debug_printf("%s New mutation boundary: '%s': %s  applyBoundaryChange=%d  boundaryExists=%d "
+				             "updatingInPlace=%d\n",
+				             context.c_str(),
+				             printable(mBegin.key()).c_str(),
+				             mBegin.mutation().toString().c_str(),
+				             applyBoundaryChange,
+				             boundaryExists,
+				             updatingInPlace);
+
 				firstMutationBoundary = false;
 
-				// Iterate over records for the mutation boundary key, keep them unless the boundary key was changed or
-				// we are not applying it
-				while (cursor.valid() && cursor.get().key == mBegin.key()) {
-					// If there were no changes to the key or we're not applying it
-					if (!applyBoundaryChange) {
-						// If not updating, add to the output set, otherwise skip ahead past the records for the
-						// mutation boundary
-						if (!updating) {
-							merged.push_back(merged.arena(), cursor.get());
-							debug_printf("%s Added %s [existing, boundary start]\n",
-							             context.c_str(),
-							             cursor.get().toString().c_str());
-						}
+				if (applyBoundaryChange) {
+					// If the boundary is being set to a value, the new KV record will be inserted
+					bool shouldInsertBoundary = mBegin.mutation().boundarySet();
+
+					// Optimization:  In-place value update of new same-sized value
+					// If the boundary exists in the page and we're in update mode and the boundary is being set to a
+					// new value of the same length as the old value then just update the value bytes.
+					if (boundaryExists && updatingInPlace && shouldInsertBoundary &&
+					    mBegin.mutation().boundaryValue.get().size() == cursor.get().value.get().size()) {
+						changesMade = true;
+						shouldInsertBoundary = false;
+
+						debug_printf("%s In-place value update for %s [existing, boundary start]\n",
+						             context.c_str(),
+						             cursor.get().toString().c_str());
+
+						copyForUpdate();
+						memcpy((uint8_t*)cursor.get().value.get().begin(),
+						       mBegin.mutation().boundaryValue.get().begin(),
+						       cursor.get().value.get().size());
 						cursor.moveNext();
-					} else {
+					} else if (boundaryExists) {
```

Comment:
Not sure if I understand the `boundaryExists` very well. Is this the delete of an existing key? Also the following comment is a bit confusing. This says, in-place update is not possible. And immediately after is the if(updatingInPlace) check.

Reply:
Ok after re-reading, it seems like updatingInPlace doesn't indicate the in-place value update. But updating the existing page? If so, it's worth clarifying.  Maybe "updating" was a more appropriate name for this variable?

Reply:
`boundaryExists` indicates that the mutation buffer boundary key exists in the DeltaTree.  Whether it must be deleted or not depends on the mutation buffer actions starting on that boundary key.

Yes `updatingInPlace` is the overall update mode of the current page.  I renamed it in an attempt to be more clear, but I agree I made it far less clear with the addition of the in-place value update.  In `updatingInPlace` the 'in place" part refers to writing a page update using the same page ID as before, vs building new page(s).

How about a rename to `updatingDeltaTree` since that's the structure actually being updated when this is set true.

Reply:
Yeah, that sounds good. In addition, adding this explanation about what it really means will be very useful.